solidity/test/CheckpointFraudProofs.t.sol

```diff
@@ -0,0 +1,271 @@
+        uint32 index = 1;
+        for (; index < checkpoints.length; index++) {
+            vm.expectRevert("message must be member of stored checkpoint");
+            cfp.isFraudulentMessageId(
+                checkpoints[index],
+                proofs[index],
+                checkpoints[index].messageId
+            );
+        }
+    }
+
+    function test_RevertWhenNotLocal_isFraudulentMessageId(
+        uint8 fixtureIndex
+    ) public {
+        vm.assume(fixtureIndex < FIXTURE_COUNT);
+
+        (
+            Checkpoint[] memory checkpoints,
+            bytes32[32][] memory proofs
+        ) = loadFixture(fixtureIndex);
+
+        for (uint32 i = 0; i < checkpoints.length; i++) {
+            checkpoints[i].origin = remoteDomain;
+            vm.expectRevert("must be local checkpoint");
+            cfp.isFraudulentMessageId(
+                checkpoints[i],
+                proofs[i],
+                checkpoints[i].messageId
+            );
+        }
+    }
+
+    function test_IsFraudulentRoot(uint8 fixtureIndex) public {
+        vm.assume(fixtureIndex < FIXTURE_COUNT);
+
+        (
+            Checkpoint[] memory checkpoints,
+            bytes32[32][] memory proofs
+        ) = loadFixture(fixtureIndex);
+
+        // cannot store checkpoint when count is 0
+        if (checkpoints.length != 0) {
+            cfp.storeLatestCheckpoint(address(merkleTreeHook));
+        }
+
+        // check all messages against latest stored checkpoint
+        for (uint32 i = 0; i < checkpoints.length; i++) {
+            assertFalse(cfp.isFraudulentRoot(checkpoints[i], proofs[i]));
+            checkpoints[i].root = ~checkpoints[i].root;
+            assertTrue(cfp.isFraudulentRoot(checkpoints[i], proofs[i]));
+        }
+    }
+
+    function test_RevertWhenNotStored_isFraudulentRoot(
+        uint8 fixtureIndex
+    ) public {
+        vm.assume(fixtureIndex < FIXTURE_COUNT);
+
+        (
+            Checkpoint[] memory checkpoints,
+            bytes32[32][] memory proofs
+        ) = loadFixture(fixtureIndex);
+
+        // stored checkpoint with index 0 is unfortunately not fraud provable
+        uint32 index = 1;
+        for (; index < checkpoints.length; index++) {
+            vm.expectRevert("message must be member of stored checkpoint");
+            cfp.isFraudulentRoot(checkpoints[index], proofs[index]);
+        }
+    }
+
+    function test_RevertWhenNotLocal_isFraudulentRoot(
+        uint8 fixtureIndex
+    ) public {
+        vm.assume(fixtureIndex < FIXTURE_COUNT);
+
+        (
+            Checkpoint[] memory checkpoints,
+            bytes32[32][] memory proofs
+        ) = loadFixture(fixtureIndex);
+
+        for (uint32 i = 0; i < checkpoints.length; i++) {
+            checkpoints[i].origin = remoteDomain;
+            vm.expectRevert("must be local checkpoint");
+            cfp.isFraudulentRoot(checkpoints[i], proofs[i]);
+        }
+    }
+}
```

Comment:
Tests for the following will be helpful:

- improper merkle proofs
- the race condition between mailbox insertion and merkle tree building

Reply:
imo neither of these are really unit tests but will add for clarity

Reply:
actually the latter would only be testable if the fraud proofs were comparing to merkle.latestCheckpoint